middle_end/flambda/basic/continuation.ml

```diff
@@ -37,58 +41,92 @@ module Sort = struct
+    | Return -> "Return"
+    | Define_root_symbol -> "Define_root_symbol"
+    | Toplevel_return -> "Toplevel_return"
+    | Exn -> "Exn"
 
-  let of_int t =
-    match t with
-    | 0 -> Normal
-    | 1 -> Return
-    | 2 -> Define_root_symbol
-    | 3 -> Toplevel_return
-    | 4 -> Exn
-    | _ -> assert false
-
-  let num_bits_needed = 3
+  let print ppf t = Format.pp_print_string ppf (to_string t)
 end
 
-let max_num_continuations = (-1) lsr Sort.num_bits_needed
-let () = assert (max_num_continuations > 0)
+module Data = struct
+  type t = {
+    compilation_unit : Compilation_unit.t;
+    name : string;
+    name_stamp : int;
+    sort : Sort.t;
+  }
+
+  let flags = continuation_flags
+
+  let print ppf { compilation_unit; name; name_stamp; sort; } =
+    Format.fprintf ppf "@[<hov 1>(\
+        @[<hov 1>(compilation_unit@ %a)@]@ \
+        @[<hov 1>(name@ %s)@]@ \
+        @[<hov 1>(name_stamp@ %d)@]@ \
+        @[<hov 1>(sort@ %a)@]\
+        )@]"
+      Compilation_unit.print compilation_unit
+      name
+      name_stamp
+      Sort.print sort
+
+  let hash { compilation_unit; name = _; name_stamp; sort = _; } =
+    Hashtbl.hash (Compilation_unit.hash compilation_unit, name_stamp)
 
-let sort_shift = Sys.int_size - Sort.num_bits_needed
+  let equal t1 t2 =
+    if t1 == t2 then true
+    else
+      let { compilation_unit = compilation_unit1; name_stamp = name_stamp1; _
+          } = t1
+      in
+      let { compilation_unit = compilation_unit2; name_stamp = name_stamp2; _
+          } = t2
+      in
+      Int.equal name_stamp1 name_stamp2
+        && Compilation_unit.equal compilation_unit1 compilation_unit2
+end
 
 (* There is no need to store the compilation unit, since in a .cmx file,
    no values of type [t] occur that are not under [Name_abstraction] binders. *)
-type t = int
+type t = Id.t
+type exported = Data.t
+
+module Table = Table_by_int_id.Make(Data)
+let grand_table_of_continuations = ref (Table.create ())
 
-let create ?sort () : t =
+let initialise () = grand_table_of_continuations := Table.create ()
+
+let create ?sort ?name () : t =
   let sort = Option.value sort ~default:Sort.Normal in
-  let id = next_raise_count () in
-  if id < max_num_continuations then
-    ((Sort.to_int sort) lsl sort_shift) lor id
-  else
-    Misc.fatal_error "Use a machine with a wider word size to compile"
+  let name = Option.value name ~default:"k" in
+  let compilation_unit = Compilation_unit.get_current_exn () in
+  let name_stamp = next_raise_count () in
+  let data : Data.t = { compilation_unit; name; name_stamp; sort } in
+  Table.add !grand_table_of_continuations data
+
+let find_data t = Table.find !grand_table_of_continuations t
+
+let rename t =
+  let { Data.name; sort; name_stamp = _; compilation_unit = _ } = find_data t in
+  create ~sort ~name ()
 
-(* CR mshinwell: Document why this uses [next_raise_count].  Does it need
-   to?  It would be better if it didn't. *)
+let name t = (find_data t).name
```

Comment:
It looks like a CR about "Document why this uses next_raise_count" has been lost at this spot, please put it back in...

Reply:
Do you know where it originally went? It seemed to already have drifted out of place and I wasn't sure where it should go.

Reply:
Just a bit further up, where that function is called, during creation.